servers/rendering/renderer_viewport.cpp

```diff
@@ -77,33 +77,17 @@ void RendererViewport::_configure_3d_render_buffers(Viewport *p_viewport) {
 			RSG::scene->free(p_viewport->render_buffers);
 			p_viewport->render_buffers = RID();
 		} else {
-			RS::ViewportScale3D scale_3d = p_viewport->scale_3d;
-			if (Engine::get_singleton()->is_editor_hint()) { // ignore this inside of the editor
-				scale_3d = RS::VIEWPORT_SCALE_3D_DISABLED;
+			float scale_3d = p_viewport->scale_3d;
+			if (Engine::get_singleton()->is_editor_hint()) {
+				// Ignore the 3D viewport render scaling inside of the editor.
+				// The Half Resolution 3D editor viewport option should be used instead.
+				scale_3d = 1.0;
 			}
 
-			int width = p_viewport->size.width;
-			int height = p_viewport->size.height;
-			switch (scale_3d) {
-				case RS::VIEWPORT_SCALE_3D_75_PERCENT: {
-					width = (width * 3) / 4;
-					height = (height * 3) / 4;
-				}; break;
-				case RS::VIEWPORT_SCALE_3D_50_PERCENT: {
-					width = width >> 1;
-					height = height >> 1;
-				}; break;
-				case RS::VIEWPORT_SCALE_3D_33_PERCENT: {
-					width = width / 3;
-					height = height / 3;
-				}; break;
-				case RS::VIEWPORT_SCALE_3D_25_PERCENT: {
-					width = width >> 2;
-					height = height >> 2;
-				}; break;
-				default:
-					break;
-			}
+			// Clamp 3D rendering resolution to reasonable values supported on most hardware.
+			// This prevents freezing the engine or outright crashing on lower-end GPUs.
+			const int width = CLAMP(p_viewport->size.width * scale_3d, 1, 16384);
```

Comment:
You should definitely not be able to scale beyond 1.0, so I would clamp scale_3d to 1.0 instead.

Reply:
Scaling beyond 1.0 makes it possible to achiece supersampling in a way that's cheaper and easier than using a viewport with a larger size. While supersampling is not very practical for games, it's very useful for offline rendering, marketing screenshots, photo modes, etc.